packages/cardano-services/src/Asset/CardanoTokenRegistry.ts

```diff
@@ -0,0 +1,141 @@
+   */
+  metadataServerUri?: string;
+}
+
+interface CardanoTokenRegistryConfigurationWithRequired extends CardanoTokenRegistryConfiguration {
+  cacheTTL: number;
+  metadataServerUri: string;
+}
+
+/**
+ * Dependencies that are need to create CardanoTokenRegistry
+ */
+export interface CardanoTokenRegistryDependencies {
+  /**
+   * The cache engine. Default: InMemoryCache with CardanoTokenRegistryConfiguration.cacheTTL as default TTL
+   */
+  cache?: InMemoryCache;
+
+  /**
+   * The logger object
+   */
+  logger: Logger;
+}
+
+/**
+ * TokenMetadataService implementation using Cardano Token Registry public API
+ */
+export class CardanoTokenRegistry implements TokenMetadataService {
+  /**
+   * The axios client used to retrieve metadata from API
+   */
+  #axiosClient: AxiosInstance;
+
+  /**
+   * The in memory cache engine
+   */
+  #cache: InMemoryCache;
+
+  /**
+   * The logger object
+   */
+  #logger: Logger;
+
+  constructor({ cache, logger }: CardanoTokenRegistryDependencies, config: CardanoTokenRegistryConfiguration = {}) {
+    const defaultConfig: CardanoTokenRegistryConfigurationWithRequired = {
+      cacheTTL: 600,
+      metadataServerUri: DEFAULT_METADATA_SERVER_URI,
+      ...config
+    };
+
+    this.#cache = cache || new InMemoryCache(defaultConfig.cacheTTL);
+    this.#axiosClient = axios.create({ baseURL: defaultConfig.metadataServerUri });
+    this.#logger = logger;
+  }
+
+  shutdown() {
+    this.#cache.shutdown();
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  #checkNumberValue(data: any, attribute: string) {
+    if (attribute in data && typeof data[attribute].value === 'number') return data[attribute].value as number;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  #checkStringValue(data: any, attribute: string) {
+    if (attribute in data && typeof data[attribute].value === 'string') return data[attribute].value as string;
+  }
+
+  async getTokenMetadata(assetIds: Cardano.AssetId[]): Promise<(Asset.TokenMetadata | null)[]> {
+    const assetIdsToRequest: Cardano.AssetId[] = [];
+    // eslint-disable-next-line unicorn/no-new-array
+    const ret: (Asset.TokenMetadata | null)[] = new Array(assetIds.length).fill(null);
+
+    this.#logger.debug(`Requested asset metatada for "${assetIds}"`);
+
+    for (const [i, assetId] of assetIds.entries()) {
+      const stringAssetId = assetId.toString();
+
+      const cachedMetadata = this.#cache.getVal<Asset.TokenMetadata>(stringAssetId);
+
+      if (cachedMetadata) {
+        this.#logger.debug(`Using chached asset metatada value for "${stringAssetId}"`);
+        ret[i] = cachedMetadata;
+      } else assetIdsToRequest.push(assetId);
+    }
+
+    // All metadata was taken from cache
+    if (assetIdsToRequest.length === 0) return ret;
+
+    this.#logger.debug(`Fetching asset metatada for "${assetIdsToRequest}"`);
+
+    try {
+      const response = await this.#axiosClient.post('metadata/query', {
```

Comment:
```suggestion
      const response = await this.#axiosClient.post<{ subjects: (Asset.TokenMetadata & { subject: string })[] }>('metadata/query', {
```
See if you can remove the custom data type coercion by providing a type upfront. I've used `Asset.TokenMetadata` as a demo, but you'll need to define the API response type given the property names vary a little.

Reply:
I found no defined `TokenMetadataServiceRecord`.
If you are suggesting it as the new type name (ok), the `Partial<>` should not be required as I can define all the properties as optional.